contracts/TokenProxy.sol

```diff
@@ -0,0 +1,150 @@
+pragma solidity ^0.4.19;
+
+contract TransferableToken {
+    function balanceOf(address _addr) public constant returns (uint256);
+    function transfer(address _to, uint256 _value) public returns (bool success);
+}
+
+
+/// The Gate is a contract with unique address to allow a token holder
+/// (called "User") to transfer tokens from original Token to the Proxy.
+///
+/// TODO: Rename to MigrationGate?
+contract Gate {
+    TransferableToken private TOKEN;
+    TokenProxy private PROXY;
+
+    /// Gates are to be created by the TokenProxy.
+    function Gate(TransferableToken _token, TokenProxy _proxy) public {
+        TOKEN = _token;
+        PROXY = _proxy;
+    }
+
+    function transferToProxy() public {
```

Comment:
Can we add a short comment that Gate contract doesn't know who is its owner, but this is checked in PROXY contract?